### PR TITLE
[WIP] Add metric to track unschedulable reasons

### LINKED ISSUE #102867: Add a metric to track reasons for unschedulable pods
#### What would you like to be added:

A metric that tracks reasons for currently unschedulable pods in a cluster. Breaking down the metric by pod name is not scalable because of cardinality, but we can break down by plugin name.

_Suggested Semantics_
```
unschedulableReasons = metrics.NewGaugeVec(		
    &metrics.GaugeOpts{
			Subsystem:      SchedulerSubsystem,
			Name:           "unschedulable_pods_by_plugin",
			Help:           "The number of unschedulable pods broken down by plugin name. A pod will increment the gauge for all plugins that caused it to not schedule and so this metric have meaning only when broken down by plugin.",
			StabilityLevel: metrics.ALPHA,
		}, []string{"plugin", "profile"})
```

Each unschedulable pod increases the gauge for all plugins that caused the pod to fail (since each pod is assessed against multiple nodes and so a pod may have one or more reasons for not being schedulable). One issue here is the gauge will have clear meaning only when broken down by plugin (number of pods unschedulable because of that plugin), but the aggregate will not have a meaning.

**_**Possible implementation:**_**
We track the plugins that caused a pod to not schedule in [QueuedPodInfo](https://github.com/kubernetes/kubernetes/blob/b766721332da82f8a3eaa7dd3f131713d291f288/pkg/scheduler/framework/types.go#L106), every time we add/remove an unschedulable pod we update the gauge.

#### Why is this needed:

This is important to setup alerts on specific reasons that may cause pods to not schedule in a cluster.

/sig scheduling

### PR DESCRIPTION
**Pull request by hprateek43**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds a gauge metric to track scheduling failure by plugin

#### Which issue(s) this PR fixes:

Fixes #102867 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/metrics/metrics.go [hunk 1]
```diff
@@ -164,6 +164,14 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		}, []string{"type"})
 
+	unschedulableReasons = metrics.NewGaugeVec(
```
Comment (alculquicondor) [inline pkg/scheduler/metrics/metrics.go:167]: CounterVec

Comment (ahg-g) [reply pkg/scheduler/metrics/metrics.go:167]: Aldo, see discussion on the main issue: https://github.com/kubernetes/kubernetes/issues/102867

Comment (alculquicondor) [reply pkg/scheduler/metrics/metrics.go:167]: Ah ok, so it's more like pending_pods metric, rather than schedule_attempts. Sounds good.

Comment (ahg-g) [reply pkg/scheduler/metrics/metrics.go:167]: Correct, we want to track the current actual number of unschedulable pods.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -568,6 +572,10 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(podInfoList []*framework.
 				klog.ErrorS(err, "Error adding pod to the scheduling queue", "pod", klog.KObj(pod))
 			} else {
 				metrics.SchedulerQueueIncomingPods.WithLabelValues("active", event.Label).Inc()
+				var val = pInfo.UnschedulablePlugins.List()
+				for _, plugin := range val {
+					metrics.UnschedulableReason(plugin, pod.Spec.SchedulerName).Dec()
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:577]: We should only decrease the metric when the Pod is getting tried, so in `scheduleOne` perhaps. But also you need to take care of deleted pods.

Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue.go:577]: Or maybe in `queue.Pop`

**Comment by ahg-g** [comment]
Lets leave this until next cycle, we may use two counters instead of a gauge.

**Pull request closed without merging**